File: TanksWinter.cs
// TanksWinter.cs — Танки (зима) на C#

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

class TanksWinter
{
    const int W = 30, H = 20;
    const char EMPTY = ' ', WALL = '#', SNOW = '.', ICE = '~';
    const char TANK = '@', ENEMY = 'E', BULLET = '*', BONUS = 'B';

    static char[,] field;
    static int playerX, playerY;
    static char playerDir = 'u';
    static List<int[]> enemies = new List<int[]>();
    static List<int[]> bullets = new List<int[]>();
    static List<int[]> bonuses = new List<int[]>();
    static int score = 0, lives = 3, level = 1, highScore = 0;
    static bool gameOver = false, paused = false;
    static Random rand = new Random();
    static DateTime lastUpdate = DateTime.Now;
    static bool quit = false;

    static void LoadHighScore() { highScore = 0; }
    static void SaveHighScore() {}

    static void InitField()
    {
        field = new char[H, W];
        for (int y=0; y<H; y++) for (int x=0; x<W; x++) field[y,x] = EMPTY;
        for (int x=0; x<W; x++) { field[0,x] = WALL; field[H-1,x] = WALL; }
        for (int y=0; y<H; y++) { field[y,0] = WALL; field[y,W-1] = WALL; }
        for (int y=2; y<H-2; y++) for (int x=2; x<W-2; x++) {
            double r = rand.NextDouble();
            if (r<0.3) field[y,x] = SNOW;
            else if (r<0.4) field[y,x] = ICE;
            else if (r<0.48) field[y,x] = WALL;
        }
    }

    static bool IsOccupied(int x, int y)
    {
        if (playerX==x && playerY==y) return true;
        foreach (var e in enemies) if (e[0]==x && e[1]==y) return true;
        foreach (var b in bullets) if (b[0]==x && b[1]==y) return true;
        return false;
    }

    static void MovePlayer(int dx, int dy)
    {
        int nx = playerX+dx, ny = playerY+dy;
        if (nx<1 || nx>=W-1 || ny<1 || ny>=H-1) return;
        if (field[ny,nx] == WALL) return;
        if (IsOccupied(nx, ny)) return;
        playerX = nx; playerY = ny;
    }

    static void ShootPlayer()
    {
        int dx=0, dy=0;
        if (playerDir=='u') dy=-1; else if (playerDir=='d') dy=1; else if (playerDir=='l') dx=-1; else dx=1;
        int bx = playerX+dx, by = playerY+dy;
        if (bx<1 || bx>=W-1 || by<1 || by>=H-1) return;
        if (field[by,bx] == WALL) { field[by,bx] = EMPTY; return; }
        bullets.Add(new int[]{bx, by, dx, dy, 1});
    }

    static void SpawnEnemy()
    {
        int side = rand.Next(4);
        int x=0, y=0;
        if (side==0) { x=rand.Next(W-4)+2; y=1; }
        else if (side==1) { x=rand.Next(W-4)+2; y=H-2; }
        else if (side==2) { x=1; y=rand.Next(H-4)+2; }
        else { x=W-2; y=rand.Next(H-4)+2; }
        if (field[y,x]!=EMPTY || IsOccupied(x,y)) return;
        enemies.Add(new int[]{x, y, 0, 0});
    }

    static void Update()
    {
        if (gameOver || paused) return;
        // Враги
        for (int i=0; i<enemies.Count; i++) {
            var e = enemies[i];
            if (rand.Next(10)<1) {
                int d = rand.Next(4);
                int dx=0, dy=0;
                if (d==0) dy=-1; else if (d==1) dy=1; else if (d==2) dx=-1; else dx=1;
                int nx=e[0]+dx, ny=e[1]+dy;
                if (nx>0 && nx<W-1 && ny>0 && ny<H-1 && field[ny,nx]!=WALL && !IsOccupied(nx, ny)) {
                    e[0]=nx; e[1]=ny;
                }
            }
            if (rand.Next(50)==0) {
                int dx=0, dy=0;
                if (e[0] < playerX) dx=1; else if (e[0] > playerX) dx=-1;
                if (e[1] < playerY) dy=1; else if (e[1] > playerY) dy=-1;
                if (dx!=0 || dy!=0) {
                    int bx=e[0]+dx, by=e[1]+dy;
                    if (bx>0 && bx<W-1 && by>0 && by<H-1 && field[by,bx]!=WALL)
                        bullets.Add(new int[]{bx, by, dx, dy, 0});
                }
            }
        }
        // Пули
        for (int i=0; i<bullets.Count; i++) {
            var b = bullets[i];
            b[0] += b[2]; b[1] += b[3];
            if (b[0]<1 || b[0]>=W-1 || b[1]<1 || b[1]>=H-1) { bullets.RemoveAt(i); i--; continue; }
            if (field[b[1],b[0]] == WALL) { field[b[1],b[0]] = EMPTY; bullets.RemoveAt(i); i--; continue; }
            bool hit = false;
            if (b[4]==0) {
                if (b[0]==playerX && b[1]==playerY) {
                    lives--; hit = true;
                    if (lives<=0) { gameOver=true; if(score>highScore) highScore=score; }
                }
            } else {
                for (int j=0; j<enemies.Count; j++) {
                    var e = enemies[j];
                    if (b[0]==e[0] && b[1]==e[1]) {
                        score++; hit = true;
                        enemies.RemoveAt(j);
                        if (score%5==0) { level++; bonuses.Add(new int[]{e[0], e[1], 'l'}); }
                        break;
                    }
                }
            }
            if (hit) { bullets.RemoveAt(i); i--; }
        }
        // Бонусы
        for (int i=0; i<bonuses.Count; i++) {
            var bon = bonuses[i];
            if (bon[0]==playerX && bon[1]==playerY) {
                if (bon[2]=='l') lives++;
                bonuses.RemoveAt(i); i--;
            }
        }
        // Спавн
        if (enemies.Count < Math.Min(3+level, 8)) {
            if (rand.Next(100) < level*1) SpawnEnemy();
        }
    }

    static void Draw()
    {
        Console.Clear();
        Console.WriteLine($"🎮 TanksWinter  |  Счёт: {score}  |  Жизни: {lives}  |  Уровень: {level}  |  Рекорд: {highScore}");
        if (paused) Console.WriteLine("⏸ ПАУЗА");
        Console.Write("+" + new string('-', W) + "+\n");
        for (int y=0; y<H; y++) {
            Console.Write("|");
            for (int x=0; x<W; x++) {
                char ch = field[y,x];
                if (x==playerX && y==playerY) ch = TANK;
                else {
                    bool found=false;
                    foreach (var e in enemies) if (e[0]==x && e[1]==y) { ch=ENEMY; found=true; break; }
                    if (!found) foreach (var b in bullets) if (b[0]==x && b[1]==y) { ch=BULLET; found=true; break; }
                    if (!found) foreach (var bon in bonuses) if (bon[0]==x && bon[1]==y) { ch=BONUS; found=true; break; }
                }
                Console.Write(ch);
            }
            Console.WriteLine("|");
        }
        Console.Write("+" + new string('-', W) + "+\n");
        Console.WriteLine("Управление: WASD - движение, Пробел - стрельба, P - пауза, Q - выход");
    }

    static int GetInput()
    {
        if (Console.KeyAvailable) {
            var key = Console.ReadKey(true).Key;
            if (key == ConsoleKey.Q) return -1;
            if (key == ConsoleKey.P) return -2;
            if (key == ConsoleKey.W) return 1;
            if (key == ConsoleKey.S) return 2;
            if (key == ConsoleKey.A) return 3;
            if (key == ConsoleKey.D) return 4;
            if (key == ConsoleKey.Spacebar) return 5;
        }
        return 0;
    }

    public static async Task Main()
    {
        LoadHighScore();
        InitField();
        playerX = W/2; playerY = H/2;
        SpawnEnemy();
        lastUpdate = DateTime.Now;
        while (!quit && !gameOver) {
            Draw();
            int inp = GetInput();
            if (inp == -1) { quit=true; break; }
            if (inp == -2) { paused = !paused; continue; }
            if (!paused) {
                if (inp == 1) { playerDir='u'; MovePlayer(0,-1); }
                else if (inp == 2) { playerDir='d'; MovePlayer(0,1); }
                else if (inp == 3) { playerDir='l'; MovePlayer(-1,0); }
                else if (inp == 4) { playerDir='r'; MovePlayer(1,0); }
                else if (inp == 5) ShootPlayer();
                if ((DateTime.Now - lastUpdate).TotalSeconds > 0.1) {
                    Update();
                    lastUpdate = DateTime.Now;
                }
            }
            await Task.Delay(20);
        }
        Console.WriteLine($"ИГРА ОКОНЧЕНА! Счёт: {score}");
        SaveHighScore();
    }
}

File: tanks_winter.rs
// tanks_winter.rs — Танки (зима) на Rust

use std::io::{self, Write, stdout};
use std::thread;
use std::time::{Duration, Instant};
use rand::Rng;
use termion::{clear, cursor, color, style};
use termion::input::TermRead;

const W: usize = 30;
const H: usize = 20;
const EMPTY: char = ' ';
const WALL: char = '#';
const SNOW: char = '.';
const ICE: char = '~';
const TANK: char = '@';
const ENEMY: char = 'E';
const BULLET: char = '*';
const BONUS: char = 'B';

#[derive(Clone)]
struct Tank {
    x: usize,
    y: usize,
    dir: char,
}

struct Bullet {
    x: usize,
    y: usize,
    dx: i32,
    dy: i32,
    player: bool,
}

struct Bonus {
    x: usize,
    y: usize,
    typ: char,
}

struct Game {
    field: Vec<Vec<char>>,
    player: Tank,
    enemies: Vec<Tank>,
    bullets: Vec<Bullet>,
    bonuses: Vec<Bonus>,
    score: i32,
    lives: i32,
    level: i32,
    high_score: i32,
    game_over: bool,
    paused: bool,
}

impl Game {
    fn new() -> Self {
        let mut game = Game {
            field: vec![vec![EMPTY; W]; H],
            player: Tank { x: W/2, y: H/2, dir: 'u' },
            enemies: Vec::new(),
            bullets: Vec::new(),
            bonuses: Vec::new(),
            score: 0,
            lives: 3,
            level: 1,
            high_score: 0,
            game_over: false,
            paused: false,
        };
        game.load_high_score();
        game.init_field();
        game.spawn_enemy();
        game
    }

    fn load_high_score(&mut self) {
        self.high_score = 0;
    }

    fn init_field(&mut self) {
        for y in 0..H {
            for x in 0..W {
                self.field[y][x] = EMPTY;
            }
        }
        for x in 0..W {
            self.field[0][x] = WALL;
            self.field[H-1][x] = WALL;
        }
        for y in 0..H {
            self.field[y][0] = WALL;
            self.field[y][W-1] = WALL;
        }
        let mut rng = rand::thread_rng();
        for y in 2..H-2 {
            for x in 2..W-2 {
                let r: f64 = rng.gen();
                if r < 0.3 {
                    self.field[y][x] = SNOW;
                } else if r < 0.4 {
                    self.field[y][x] = ICE;
                } else if r < 0.48 {
                    self.field[y][x] = WALL;
                }
            }
        }
    }

    fn is_occupied(&self, x: usize, y: usize) -> bool {
        if self.player.x == x && self.player.y == y { return true; }
        for e in &self.enemies { if e.x == x && e.y == y { return true; } }
        for b in &self.bullets { if b.x == x && b.y == y { return true; } }
        false
    }

    fn move_tank(&mut self, tank: &mut Tank, dx: i32, dy: i32) -> bool {
        let nx = tank.x as i32 + dx;
        let ny = tank.y as i32 + dy;
        if nx < 1 || nx >= W as i32 -1 || ny < 1 || ny >= H as i32 -1 { return false; }
        let nxu = nx as usize; let nyu = ny as usize;
        if self.field[nyu][nxu] == WALL { return false; }
        if self.is_occupied(nxu, nyu) { return false; }
        tank.x = nxu; tank.y = nyu;
        true
    }

    fn shoot_player(&mut self) {
        let (dx, dy) = match self.player.dir {
            'u' => (0, -1),
            'd' => (0, 1),
            'l' => (-1, 0),
            'r' => (1, 0),
            _ => (0,0),
        };
        let bx = self.player.x as i32 + dx;
        let by = self.player.y as i32 + dy;
        if bx < 1 || bx >= W as i32 -1 || by < 1 || by >= H as i32 -1 { return; }
        let bxu = bx as usize; let byu = by as usize;
        if self.field[byu][bxu] == WALL {
            self.field[byu][bxu] = EMPTY;
            return;
        }
        self.bullets.push(Bullet { x: bxu, y: byu, dx, dy, player: true });
    }

    fn spawn_enemy(&mut self) {
        let mut rng = rand::thread_rng();
        let side = rng.gen_range(0..4);
        let (x, y) = match side {
            0 => (rng.gen_range(2..W-2), 1),
            1 => (rng.gen_range(2..W-2), H-2),
            2 => (1, rng.gen_range(2..H-2)),
            _ => (W-2, rng.gen_range(2..H-2)),
        };
        if self.field[y][x] != EMPTY || self.is_occupied(x, y) { return; }
        self.enemies.push(Tank { x, y, dir: 'u' });
    }

    fn update(&mut self) {
        if self.game_over || self.paused { return; }
        let mut rng = rand::thread_rng();
        // Враги
        for i in 0..self.enemies.len() {
            if rng.gen_range(0..10) < 1 {
                let dirs = ['u','d','l','r'];
                let d = dirs[rng.gen_range(0..4)];
                let (dx, dy) = match d {
                    'u' => (0, -1),
                    'd' => (0, 1),
                    'l' => (-1, 0),
                    'r' => (1, 0),
                    _ => (0,0),
                };
                self.move_tank(&mut self.enemies[i], dx, dy);
            }
            if rng.gen_range(0..50) == 0 {
                // стрельба врага
                let mut dx = 0; let mut dy = 0;
                if self.enemies[i].x < self.player.x { dx = 1; }
                else if self.enemies[i].x > self.player.x { dx = -1; }
                if self.enemies[i].y < self.player.y { dy = 1; }
                else if self.enemies[i].y > self.player.y { dy = -1; }
                if dx != 0 || dy != 0 {
                    let bx = self.enemies[i].x as i32 + dx;
                    let by = self.enemies[i].y as i32 + dy;
                    if bx > 0 && bx < W as i32 -1 && by > 0 && by < H as i32 -1 {
                        let bxu = bx as usize; let byu = by as usize;
                        if self.field[byu][bxu] != WALL {
                            self.bullets.push(Bullet { x: bxu, y: byu, dx, dy, player: false });
                        }
                    }
                }
            }
        }
        // Пули
        let mut i = 0;
        while i < self.bullets.len() {
            let b = &mut self.bullets[i];
            b.x = (b.x as i32 + b.dx) as usize;
            b.y = (b.y as i32 + b.dy) as usize;
            if b.x < 1 || b.x >= W-1 || b.y < 1 || b.y >= H-1 {
                self.bullets.remove(i);
                continue;
            }
            if self.field[b.y][b.x] == WALL {
                self.field[b.y][b.x] = EMPTY;
                self.bullets.remove(i);
                continue;
            }
            let mut hit = false;
            if !b.player {
                if b.x == self.player.x && b.y == self.player.y {
                    self.lives -= 1;
                    hit = true;
                    if self.lives <= 0 {
                        self.game_over = true;
                        if self.score > self.high_score { self.high_score = self.score; }
                    }
                }
            } else {
                let mut j = 0;
                while j < self.enemies.len() {
                    if b.x == self.enemies[j].x && b.y == self.enemies[j].y {
                        self.score += 1;
                        hit = true;
                        self.enemies.remove(j);
                        if self.score % 5 == 0 {
                            self.level += 1;
                            self.bonuses.push(Bonus { x: b.x, y: b.y, typ: 'l' });
                        }
                        break;
                    }
                    j += 1;
                }
            }
            if hit {
                self.bullets.remove(i);
            } else {
                i += 1;
            }
        }
        // Бонусы
        let mut i = 0;
        while i < self.bonuses.len() {
            if self.bonuses[i].x == self.player.x && self.bonuses[i].y == self.player.y {
                if self.bonuses[i].typ == 'l' { self.lives += 1; }
                self.bonuses.remove(i);
            } else {
                i += 1;
            }
        }
        // Спавн врагов
        if self.enemies.len() < (3 + self.level as usize).min(8) {
            if rng.gen_range(0..100) < self.level * 1 {
                self.spawn_enemy();
            }
        }
    }

    fn draw(&self) {
        print!("{}{}", clear::All, cursor::Goto(1,1));
        println!("🎮 TanksWinter  |  Счёт: {}  |  Жизни: {}  |  Уровень: {}  |  Рекорд: {}", self.score, self.lives, self.level, self.high_score);
        if self.paused { println!("⏸ ПАУЗА"); }
        print!("+");
        for _ in 0..W { print!("-"); }
        println!("+");
        for y in 0..H {
            print!("|");
            for x in 0..W {
                let mut ch = self.field[y][x];
                if x == self.player.x && y == self.player.y {
                    ch = TANK;
                } else {
                    let mut found = false;
                    for e in &self.enemies {
                        if e.x == x && e.y == y { ch = ENEMY; found = true; break; }
                    }
                    if !found {
                        for b in &self.bullets {
                            if b.x == x && b.y == y { ch = BULLET; found = true; break; }
                        }
                    }
                    if !found {
                        for bon in &self.bonuses {
                            if bon.x == x && bon.y == y { ch = BONUS; break; }
                        }
                    }
                }
                print!("{}", ch);
            }
            println!("|");
        }
        print!("+");
        for _ in 0..W { print!("-"); }
        println!("+");
        println!("Управление: WASD - движение, Пробел - стрельба, P - пауза, Q - выход");
        stdout().flush().unwrap();
    }

    fn run(&mut self) {
        let stdin = io::stdin();
        let mut keys = stdin.keys();
        let mut last_update = Instant::now();
        while !self.game_over {
            self.draw();
            if let Some(Ok(key)) = keys.next() {
                match key {
                    termion::event::Key::Char('q') => break,
                    termion::event::Key::Char('p') => self.paused = !self.paused,
                    termion::event::Key::Char('w') => { self.player.dir = 'u'; self.move_tank(&mut self.player, 0, -1); }
                    termion::event::Key::Char('s') => { self.player.dir = 'd'; self.move_tank(&mut self.player, 0, 1); }
                    termion::event::Key::Char('a') => { self.player.dir = 'l'; self.move_tank(&mut self.player, -1, 0); }
                    termion::event::Key::Char('d') => { self.player.dir = 'r'; self.move_tank(&mut self.player, 1, 0); }
                    termion::event::Key::Char(' ') => self.shoot_player(),
                    _ => {}
                }
            }
            if last_update.elapsed().as_secs_f64() > 0.1 {
                self.update();
                last_update = Instant::now();
            }
            thread::sleep(Duration::from_millis(20));
        }
        println!("ИГРА ОКОНЧЕНА! Счёт: {}", self.score);
    }
}

fn main() {
    let mut game = Game::new();
    game.run();
}
